src/hub.js

```diff
@@ -117,79 +106,122 @@ export function hub(chrome) {
 			);
 		} else {
 			debug('App timed out, considering it dead! %o', appId);
-			call('timeout', [targetWindow, appId]);
+			handleAppTimeout(appId, targetWindow);
 			try {
-				call('kill', [targetWindow]);
+				forgetApp(targetWindow);
 			} catch (error) {
 				console.error(error);
 			}
 		}
 	}
 
+	function handleAppConnect({ data: message, source: sourceWindow }) {
+		const appId = appByWindow(sourceWindow);
+		const token = uuid();
+		const spawnWaiting = appSpawns.findIndex(spawn => spawn.appId === appId);
+		const connackMessage = {
+			type: 'CONNACK',
+			viaHub: true,
+			target: appId,
+			token
+		};
+		appWindows.set(sourceWindow, { appId, token });
+
+		debug('CONNECT %o', appId);
+
+		if (spawnWaiting !== -1) {
+			const { data, source } = appSpawns[spawnWaiting].message;
+			connackMessage.source = source;
+			connackMessage.data = data;
+			appSpawns.splice(spawnWaiting, 1);
+		}
+
+		postMessage(connackMessage, sourceWindow);
+
+		const pingOpts = { targetWindow: sourceWindow, appId, token };
+		const timeoutId = setTimeout(
+			() => pingApp(pingOpts),
+			hubInstance.HEARTBEAT
+		);
+
+		appPongs.set(token, {
+			lastPong: window.performance.now(),
+			timeoutIds: new Set([timeoutId])
+		});
+	}
+
+	function handleSpawn({ data: message, source: sourceWindow }) {
+		debug('Spawning %o from %o - %O', message.target, message.source, message);
+		try {
+			const appId = handleAppSpawn(message.target, message.source);
+			appSpawns.push({ appId, message });
+		} catch (e) {
+			postMessage({
+				type: 'SPAWN-FAIL',
```

Comment:
`TODO` for later:

we should move all the message/event types to constants, in case of weird typos or if we want to namespace them.